### PR TITLE
Add an option to let the music play continuously without restart if…

### PR DESCRIPTION
**Pull request by CommonLoon102**
…the player loads a game with the same map the player is already playing on.

On higher difficulties it is often very annoying that the music restarts from the beginning every time the player loads a saved game for the same map. It causes that only the first few seconds are played over and over again. In case of CD audio it is even worse because some of the tracks are starting with silence or low volume.

The original game had this behavior too, so I've kept it as the default, and I've added a new audio setting for this feature. BloodGDX has this feature too but it is enabled by default and cannot be turned off. My philosophy is that every new feature or modification should be opt-in to keep the game as close to the original as  possible.

Related issue: https://github.com/nukeykt/NBlood/issues/26

### CODE REVIEW

**Comment by CommonLoon102** [comment]
Do not merge this yet, I found some bugs related to changing maps, in that case no music is played, I'll have to work on this a little more.

**Comment by CommonLoon102** [comment]
E1M5 tricked me as it doesn't have any music defined in the blood.ini, so it was OK. But I found another bug: save the game before ending a level, end the level and go to the next, load the save. In this case the second level's music was played instead of the first one. This is now fixed. Also, now if a demo was played the game will restart the music anyways, even if the loaded map is the same what was played in the demo.

File: source/blood/src/blood.cpp [hunk 1]
```diff
@@ -463,7 +466,8 @@ void PreloadCache(void)
     char tempbuf[128];
     if (gDemo.at1)
         return;
-    sndPlaySpecialMusicOrNothing(MUS_LOADING);
+    if (MusicRestartsOnLoadToggle)
+        sndPlaySpecialMusicOrNothing(MUS_LOADING);
```
Comment (nukeykt) [inline source/blood/src/blood.cpp]: I think you can use sndTryPlaySpecialMusic here instead

File: source/blood/src/config.cpp [hunk 2]
```diff
@@ -322,6 +323,7 @@ void CONFIG_SetDefaults(void)
     MouseBias       = 0;
     MouseDeadZone   = 0;
     MusicToggle     = 1;
+    MusicRestartsOnLoadToggle = 1;
```
Comment (nukeykt) [inline source/blood/src/config.cpp]: Default it to 0

File: source/blood/src/loadsave.cpp [hunk 3]
```diff
@@ -91,7 +91,7 @@ void LoadSave::Write(void *pData, int nSize)
         ThrowError("File error #%d writing save file.", errno);
 }
 
-void LoadSave::LoadGame(char *pzFile)
+void LoadSave::LoadGame(char *pzFile, bool demoWasPlayed)
```
Comment (nukeykt) [inline source/blood/src/loadsave.cpp]: You could move `gDemo.at1` and `gDemo.Close()` here from `void LoadGame(CGameMenuItemZEditBitmap *pItem, CGameMenuEvent *event)` instead

File: source/blood/src/menu.cpp [hunk 4]
```diff
@@ -569,8 +570,9 @@ const char *pzMusicDeviceStrings[] = {
 };
 
 CGameMenuItemTitle itemOptionsSoundTitle("SOUND SETUP", 1, 160, 20, 2038);
-CGameMenuItemZBool itemOptionsSoundSoundToggle("SOUND:", 3, 66, 60, 180, false, UpdateSoundToggle, NULL, NULL);
-CGameMenuItemZBool itemOptionsSoundMusicToggle("MUSIC:", 3, 66, 70, 180, false, UpdateMusicToggle, NULL, NULL);
+CGameMenuItemZBool itemOptionsSoundSoundToggle("SOUND:", 3, 66, 50, 180, false, UpdateSoundToggle, NULL, NULL);
+CGameMenuItemZBool itemOptionsSoundMusicToggle("MUSIC:", 3, 66, 60, 180, false, UpdateMusicToggle, NULL, NULL);
+CGameMenuItemZBool itemOptionsSoundMusicRestartsOnLoadToggle("MUSIC RESTARTS:", 3, 66, 70, 180, false, UpdateMusicRestartsOnLoadToggle, NULL, NULL);
```
Comment (nukeykt) [inline source/blood/src/menu.cpp]: I don't think this option is worth exposing to menu

File: source/blood/src/config.cpp (hunk 2 above)
Comment (CommonLoon102) [reply source/blood/src/config.cpp]: That would be different from the original game, are you sure you want to make behavioral changes by default compared the the original? Are there already things what differs from the original on purpose already?

File: source/blood/src/menu.cpp (hunk 4 above)
Comment (CommonLoon102) [reply source/blood/src/menu.cpp]: If you are sure that you want to change by default the original behavior then I agree, the console command would be enough for 99.99% of the players.

File: source/blood/src/config.cpp (hunk 2 above)
Comment (nukeykt) [reply source/blood/src/config.cpp]: i'm fine with it, this is not chocolate port anyway. In long run it'll be EDuke32 equivalent for blood

**Comment by nukeykt** [comment]
Thanks

**Pull request merged**